File: matrix/row.go
package matrix

import (
	"fmt"
)

type Row []*Cell

// Lhs returns the left-hand-side of the matrix.
func (r Row) Lhs() Row {
	return r[:len(r)-1]
}

// Rhs returns the right-hand-side of the matrix.
func (r Row) Rhs() *Cell {
	return r[len(r)-1]
}

// ConstrainedLhs returns the non-zero left-hand-side cells of the matrix.
func (r Row) ConstrainedLhs() Row {
	var row Row
	for _, cell := range r.Lhs() {
		if cell.value != 0 {
			row = append(row, cell)
		}
	}
	return row
}

// validate validates that the left-hand-cells in the row are consistent with the
// right-hand-side.
func (r Row) validate() error {
	lhsNegativeSum := 0.0
	lhsPositiveSum := 0.0
	for _, cell := range r.ConstrainedLhs() {
		if cell.value < 0 {
			lhsNegativeSum += cell.value
			continue
		}
		lhsPositiveSum += cell.value
	}
	if r.Rhs().value > lhsPositiveSum || r.Rhs().value < lhsNegativeSum {
		return fmt.Errorf("invalid constraint for row %d", r.Rhs().space.Index())
	}
	return nil
}

File: src/matrix/row.rs
use super::cell::Cell;

pub struct Row {
    pub cells: Vec<Cell>,
}

impl Row {
    pub fn new(cells: Vec<Cell>) -> Self {
        Self { cells }
    }

    pub fn lhs(&self) -> &[Cell] {
        &self.cells[..self.cells.len() - 1]
    }

    pub fn rhs(&self) -> &Cell {
        &self.cells[self.cells.len() - 1]
    }

    pub fn rhs_mut(&mut self) -> &mut Cell {
        let len = self.cells.len();
        &mut self.cells[len - 1]
    }

    pub fn constrained_lhs(&self) -> Vec<&Cell> {
        self.lhs()
            .iter()
            .filter(|cell| cell.value != 0.0)
            .collect()
    }

    pub fn validate(&self) -> Result<(), String> {
        let mut lhs_negative_sum = 0.0;
        let mut lhs_positive_sum = 0.0;

        for cell in self.constrained_lhs() {
            if cell.value < 0.0 {
                lhs_negative_sum += cell.value;
            } else {
                lhs_positive_sum += cell.value;
            }
        }

        let rhs_value = self.rhs().value;
        if rhs_value > lhs_positive_sum || rhs_value < lhs_negative_sum {
            return Err(format!(
                "invalid constraint for row {}",
                self.rhs().space_idx.get()
            ));
        }

        Ok(())
    }
}
